what4/src/Test/Verification.hs

```diff
@@ -0,0 +1,194 @@
+  , getSize
+
+    -- * Test concretization
+    --
+    -- Used by test implementation functions to map from this
+    -- Verification abstraction to the actual test mechanism
+    -- (e.g. QuickCheck, HedgeHog, etc.)
+  , Assumption(..)
+  , GenEnv(..)
+  , toNativeProperty
+  )
+where
+
+import Control.Monad.Trans.Reader
+
+-- | Local definition of a Property: intended to be a proxy for a
+-- QuickCheck Property or a Hedgehog Property.  The 'toNativeProperty'
+-- implementation function converts from these proxy Properties to the
+-- native Property implementation.
+--
+-- Tests should only use the 'Property' type as an output; the
+-- constructors and internals should be used only by the test
+-- concretization.
+data Property = BoolProperty Bool
+              | AssumptionProp Assumption
+              | PropProperty Property
+  deriving Show
+
+-- | A class specifying things that can be verified by constructing a
+-- local Property.
+class Verifiable prop where
+  verifying :: prop -> Property
+
+instance Verifiable Bool where verifying = BoolProperty
+
+-- | Used by testing code to assert a boolean property.
+property :: Bool -> Property
+property = verifying
+
+-- | Internal data structure to store the two elements to the '==>'
+-- assumption operator.
+data Assumption  = Assuming { preCondition :: Bool,
+                              assumedProp :: Property }
+  deriving Show
+
+
+-- | The named form of the '==>' assumption operator
+assuming :: Verifiable t => Bool -> t -> Property
+assuming precond test = AssumptionProp $ Assuming precond $ verifying test
+
+-- | The assumption operator that performs the property test (second
+-- element) only when the first argument is true (the assumption guard
+-- for the test).  This is the analog to the corresponding QuickCheck
+-- ==> operator.
+(==>) :: Verifiable t => Bool -> t -> Property
+(==>) = assuming
+infixr 0 ==>
+
+
+instance Verifiable Property where
+  verifying = PropProperty
+
+-- ----------------------------------------------------------------------
+
+-- | This is the reader environment for the surface level proxy
+-- testing monad.  This environment will be provided by the actual
+-- test code to map these proxy operations to the specific testing
+-- implementation.
+data GenEnv m = GenEnv { genChooseBool :: m Bool
+                       , genChooseInt :: (Int, Int) -> m Int
+                       , genChooseInteger :: (Integer, Integer) -> m Integer
+                       , genGetSize :: m Int
+                       }
+
+-- | This is the generator monad for the Verification proxy tests.
+-- The inner monad will be the actual test implementation's monadic
+-- generator, and the 'a' return type is the type returned by running
+-- this monad.
+--
+-- Tests should only use the 'GenV m TYPE' as an output; the
+-- constructors and internals should be used only by the test
+-- concretization.
+newtype GenV m a = GenV { unGenV :: ReaderT (GenEnv m) m a }
```

Comment:
Can we go even further?  What about:
```
data GenV a = GenV (forall m. Monad m => ReaderT (GenEnv m) m a)
```

Reply:
Interesting: that would reduce the impact on test sites.  Renaming the type from `GenV` back to `Gen` would make it almost a no-change drop-in replacement.